File: prototype/rails/src/app/models/page.rb
# Which slice of a listing grid a visitor is looking at, and the two numbers
# a query needs to fetch it.
Page = Data.define(:number, :size, :total_count) do
  # +requested+ arrives from a query string, so anything that is not a page
  # of this collection lands on the nearest one that is.
  def self.of(requested:, size:, total_count:)
    raise ArgumentError, "a page holds at least one item, got #{size}" unless size.is_a?(Integer) && size.positive?
    raise ArgumentError, "a count cannot be negative, got #{total_count}" if total_count.negative?

    last = last_number(size: size, total_count: total_count)

    new(number: requested.to_i.clamp(1, last), size: size, total_count: total_count)
  end

  # An empty collection still has a first page, which is where "no art
  # matches that" is written.
  def self.last_number(size:, total_count:)
    [(total_count + size - 1) / size, 1].max
  end

  def offset
    (number - 1) * size
  end

  def limit
    size
  end

  def count
    Page.last_number(size: size, total_count: total_count)
  end

  def first?
    number == 1
  end

  def last?
    number == count
  end

  def previous_number
    number - 1
  end

  def next_number
    number + 1
  end
end

File: prototype/php/src/app/Support/Page.php
<?php

declare(strict_types=1);

namespace App\Support;

use InvalidArgumentException;

/**
 * Which slice of a total-count-based list a viewer is looking at, and the
 * offset/limit a query needs to fetch it. `of()` clamps whatever a query
 * string asked for onto the nearest real page, the way every other admin
 * filter treats an out-of-range request as "give me the closest sane thing"
 * rather than a 400 — unlike the filter values `LogsQueryRequest` gates.
 */
final readonly class Page
{
    public function __construct(
        public int $number,
        public int $size,
        public int $totalCount,
        public int $count,
        public int $offset,
        public int $limit,
        public bool $isFirst,
        public bool $isLast,
        public int $previousNumber,
        public int $nextNumber,
    ) {}

    public static function of(?string $requested, int $size, int $totalCount): self
    {
        if ($size < 1) {
            throw new InvalidArgumentException("a page holds at least one item, got {$size}");
        }

        if ($totalCount < 0) {
            throw new InvalidArgumentException("a count cannot be negative, got {$totalCount}");
        }

        $count = max((int) ceil($totalCount / $size), 1);
        $asked = $requested !== null && is_numeric($requested) ? (int) $requested : 1;
        $number = self::clamp($asked, 1, $count);

        return new self(
            number: $number,
            size: $size,
            totalCount: $totalCount,
            count: $count,
            offset: ($number - 1) * $size,
            limit: $size,
            isFirst: $number === 1,
            isLast: $number === $count,
            previousNumber: $number - 1,
            nextNumber: $number + 1,
        );
    }

    private static function clamp(int $value, int $low, int $high): int
    {
        return min(max($value, $low), $high);
    }
}
